llms/mlx_lm/server.py

```diff
@@ -51,6 +51,19 @@ def stopping_criteria(
     return StopCondition(stop_met=False, trim_length=0)
 
 
+def openai_to_common_denominator(messages: List[dict]):
+    # fuse system and user, and strictly alternate with roles
+    for m in messages:
+        m["role"] = m["role"].replace("system", "user")
+    t_messages = []
+    for m in messages:
+        if t_messages == [] or t_messages[-1]["role"] != m["role"]:
+            t_messages.append(m)
+        else:
+            t_messages[-1]["content"] = "\n" + m["content"]
+    return t_messages
```

Comment:
This chat template stuff makes my head spin a little. But shouldn't this be handled on the client side? It's a bit odd to me to hardcode the prompt modification on the server when the client can specify whatever they want in the prompt.

Reply:
I agree it's not great. But my use case is precisely to use the OpenAI server to serve MLX LLMs when I have no control over the client, e.g. using an app or a library that supports the OpenAI API but not MLX out of the box. I am OK if you think this is out of scope. Thanks and sorry for the head spin :)

Reply:
I just have a question about that. Is the reason you have to do the mapping because other LLMs have different chat templates than what OpenAI uses? So the client side calls MLX server as if it's OpenAI but on the backend the model expects a different template?

But in that case this is pretty specific to a given LM right? Which LMs does it work for?

I'm just thinking about how to handle this in a more flexible way.

Reply:
Yes, that's exactly the reason... It seems pretty specific to the LM, but trying many of them, it seems the common denominator that they can all accept is alternating user/assistant messages.